pep-0724.rst

```diff
@@ -0,0 +1,235 @@
+PEP: 724
+Title: Stricter Type Guards
+Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <erictr at microsoft.com>, Erik De Bonte <erikd at microsoft.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 28-Jul-2023
+Python-Version: 3.13
+Post-History: 
+Resolution: 
+
+
+Abstract
+========
+
+:pep:`647` introduced the concept of ``TypeGuard`` functions which return true
+if their input parameter matches their target type. For example, a function that
+returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's
+input parameter is a ``str``. This allows type checkers to narrow types in this
+positive case.
+
+This PEP further refines :pep:`647` by allowing type checkers to also narrow types
+when a ``TypeGuard`` function returns false.
+
+Motivation
+==========
+
+``TypeGuards`` are used throughout Python libraries to allow a type checker to
+narrow the type of something when the ``TypeGuard`` returns true.
+
+However, in the ``else`` clause, :pep:`647` didn't prescribe what the type might
+be:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type here is not narrowed. It is still 'str | int' 
+
+
+This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype
+of the type of the first input parameter, then the false return case can be
+further narrowed. 
+
+This changes the example above like so:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type checkers can assume val is an 'int' in this branch
+
+Since the ``TypeGuard`` type (or output type) is a subtype of the first input
+parameter type, a type checker can determine that the only possible type in the
+``else`` is the other type in the ``Union``. In this example, it is safe to
+assume that if ``is_str`` returns false, then type of the ``val`` argument is an
+``int``.
+
+
+Specification
```

Comment:
The spec talks about type narrowing in the negative case (which is part of what "strict type guard semantics" provide), but it doesn't talk about narrowing of unions based on constraints imposed by the type guard return type (which is the second part of what strict type guard semantics provide). For details, refer to my original post [here](https://github.com/python/typing/discussions/1013#discussioncomment-1966238). In particular, pay close attention to point 3.

I recommend that you split the specification into two part:
1. The conditions under which "strict type guard semantics" will be applied. This is effectively point 1 from my original post (but instead of enforcing the requirement, it uses this as a condition as a test for whether strict semantics can be applied).
2. A detailed description of what is implied by "strict type guard semantics". This should include both points 2 and 3 from my original post.